src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -1174,6 +1174,41 @@ def _poly(self, i):
             return binomial(x, i)
 
         class Element(CombinatorialFreeModule.Element):
-            pass
+            def variable_shift(self, k=1):
+                r"""
+                Return the image by the shift of variables.
+
+                On polynomials, the action is the shift
+                on variables `x \mapsto x + k`.
+
+                INPUT:
+
+                - `k` -- integer (default: 1)
+
+                EXAMPLES::
+
+                    sage: A = IntegerValuedPolynomialRing(ZZ).B()
+                    sage: B = A.basis()
+                    sage: B[5].variable_shift()
+                    B[4] + B[5]
+                    sage: B[5].variable_shift(-1)
+                    -B[0] + B[1] - B[2] + B[3] - B[4] + B[5]
+
+                TESTS::
+
+                    sage: B[5].variable_shift(0)
+                    B[5]
+                    sage: B[5].variable_shift().variable_shift(-1)
+                    B[5]
+                """
+                if k == 0:
+                    return self
+                A = self.parent()
+
+                def on_basis(n):
+                    return A._from_dict({A._indices(j): binomial(k, n - j)
+                                         for j in range(n + 1)})
+
+                return A.module_morphism(on_basis, codomain=A)(self)
```

Comment:
```suggestion
                def on_basis(n):
                    return {A._indices(j): binomial(k, n - j) for j in range(n + 1)}

                from sage.data_structures.blas_dict import linear_combination
                mc = self._monomial_coefficients
                ret = linear_combination((on_basis((index), coeff)
                                         for (index, coeff) in mc.items())
                return A.element_class(A, ret)
```
This should be faster as it creates far fewer transient objects (zeros have also already been removed).